### PR TITLE
Refactor the `WriteConfig` function for bufconfig for a generic use case

### PR DESCRIPTION
**Pull request by doriable**
Add `WriteConfigWithBreakingConfig`, `WriteConfigWithLintConfig`, and `WriteConfigWithVersion` options for the `bufconfig.WriteConfig` function to support more generic use cases of the `WriteConfig`.
Before this change, we are only using `bufconfig.WriteConfig` in the `buf config init` command, but we now need it to write configuration structures into our module cache so that we are persisting `breaking` and `lint` config in our cache for our module structure.

### CODE REVIEW

File: private/bufpkg/bufconfig/write.go [hunk 1]
```diff
@@ -273,30 +257,162 @@ func writeConfig(
...
-	if writeConfigOptions.moduleIdentity == nil && len(writeConfigOptions.dependencyModuleReferences) > 0 {
-		return errors.New("cannot set deps without a name for WriteConfig")
+	// This is the same default as the bufconfig getters.
+	version := V1Beta1Version
+	if writeConfigOptions.version != "" {
+		if err := validateVersion(writeConfigOptions.version); err != nil {
+			return err
+		}
+		version = writeConfigOptions.version
 	}
```
Comment (amckinney) [inline private/bufpkg/bufconfig/write.go:270]: Is this right? The default version is `v1` when we run `buf config init`, and the `WriteConfigWithVersion` option says `v1` is still the default, but this looks like `v1beta1` is.

I've tested that `buf config init` continues to write a configuration with `v1` on this branch, so I guess it's working as expected, but I wonder if it's just because we're always providing the options at each call-site.

File: private/buf/cmd/buf/buf_test.go [hunk 2]
```diff
@@ -1158,12 +1158,12 @@ func TestConfigInitBasic(t *testing.T) {
...
-  use:
-    - DEFAULT
 breaking:
-  use:
-    - FILE
+    use:
+        - FILE
+lint:
+    use:
+        - DEFAULT
```
Comment (amckinney) [inline private/buf/cmd/buf/buf_test.go:1166]: Gah yea this isn't great, especially because it effects how users interact with `buf config init`. Unless there's a `yaml.Marshal` option we can take advantage of to control indentation, this might require that the `buf config init` path takes on a different path than the configuration written to the module cache.

Unfortunately, the aesthetic of the configuration we write with `buf config init` matters here.

Check out https://pkg.go.dev/gopkg.in/yaml.v3#Encoder.SetIndent and see if we can get that to work.

File: private/bufpkg/bufconfig/write.go [hunk 3]
```diff
@@ -273,30 +257,162 @@ func writeConfig(
...
-	for _, dependencyModuleReference := range writeConfigOptions.dependencyModuleReferences {
-		externalConfigV1.Deps = append(
-			externalConfigV1.Deps,
-			dependencyModuleReference.String(),
-		)
+	if writeConfigOptions.lintConfig != nil {
+		config.Lint = writeConfigOptions.lintConfig
+		// Always force the provided version.
+		config.Lint.Version = version
```
Comment (amckinney) [inline private/bufpkg/bufconfig/write.go:283]: We should validate that the provided `lintConfig.Version` is equivalent to the `version` (same with `breakingConfig` above).

Forcing the override might lead to invalid results if the configured rules changed between `v1beta` and `v1`.

File: private/bufpkg/bufconfig/bufconfig.go [hunk 4]
```diff
@@ -143,6 +143,35 @@ func WriteConfigWithDependencyModuleReferences(dependencyModuleReferences ...buf
 	}
 }
 
+// WriteConfigWithBreakingConfig returns a new WriteConfigOption that sets the breaking change
+// config for the module.
+//
+// If this option is used and the version used will be consistent with the rest of the config.
```
Comment (amckinney) [inline private/bufpkg/bufconfig/bufconfig.go:149]: nit: If we update the behavior to validate, let's update the wording to match it.
```suggestion
// If this option is used then the version used must be consistent with the rest of the config.
```

File: private/bufpkg/bufconfig/bufconfig.go [hunk 5]
```diff
@@ -143,6 +143,35 @@ func WriteConfigWithDependencyModuleReferences(dependencyModuleReferences ...buf
...
+func WriteConfigWithBreakingConfig(breakingConfig *bufbreakingconfig.Config) WriteConfigOption {
+	return func(writeConfigOptions *writeConfigOptions) {
+		writeConfigOptions.breakingConfig = breakingConfig
+	}
+}
+
+// WriteConfigWithLintConfig returns a new WriteConfigOption that sets the lint config for // the module.
+//
+// If this option is used and the version used will be consistent with the rest of the config.
```
Comment (amckinney) [inline private/bufpkg/bufconfig/bufconfig.go:158]: nit: Similar to above
```suggestion
// If this option is used then the version used must be consistent with the rest of the config.
```

File: private/bufpkg/bufconfig/bufconfig.go [hunk 6]
```diff
@@ -143,6 +143,35 @@ func WriteConfigWithDependencyModuleReferences(dependencyModuleReferences ...buf
...
+func WriteConfigWithLintConfig(lintConfig *buflintconfig.Config) WriteConfigOption {
+	return func(writeConfigOptions *writeConfigOptions) {
+		writeConfigOptions.lintConfig = lintConfig
+	}
+}
+
+// WriteConfigWithVersion returns a new WriteConfigOption that sets the version of the config
+// being written.
+//
+// If this is not set, the default is v1.
```
Comment (amckinney) [inline private/bufpkg/bufconfig/bufconfig.go:168]: Depending on the conversation below, the default might be `v1beta1` to preserve the default value of `buf.yaml`, but I'm not totally sure yet if this option will match that behavior from a `WriteConfig` standpoint.

File: private/bufpkg/bufconfig/write.go [hunk 7]
```diff
@@ -273,30 +257,162 @@ func writeConfig(
...
+	config *Config,
+	dependencies []string,
+) error {
+	var name string
+	if config.ModuleIdentity != nil {
+		name = config.ModuleIdentity.IdentityString()
+	}
+	breakingConfig := config.Breaking
+	lintConfig := config.Lint
+	if version == V1Beta1Version {
```
Comment (amckinney) [inline private/bufpkg/bufconfig/write.go:313]: We should still write this so that the `version` must be either `v1beta1` or `v1` (instead of defaulting to `v1` when version is not `v1beta1`). I know we already validate it via `validateVersion` earlier, but this implicit default behavior is a little less obvious.

For consistency, we can adopt the `switch` approach that you see in the other configuration functions, like [getConfigForDataInternal](https://github.com/bufbuild/buf/blob/e4a9b7b053022398f70922a9d7241b822154fe5d/private/bufpkg/bufconfig/get.go#L88).

Similarly, we could write separate helper functions (e.g. `externalConfig{V1,V1Beta1}ForConfig`) so that the implementation reads similar to the `case` found in `getConfigForDataInternal`:

```go
case V1Beta1Version:
  var externalConfigV1Beta1 ExternalConfigV1Beta1
  if err := unmarshalStrict(data, &externalConfigV1Beta1); err != nil {
    return nil, err
  }
  return newConfigV1Beta1(externalConfigV1Beta1)
```

File: private/bufpkg/bufmodule/bufmodule.go [hunk 8]
```diff
@@ -472,7 +473,37 @@ func ModuleToBucket(
...
+	if breakingConfigVersion != "" && lintConfigVersion != "" {
+		if breakingConfigVersion != lintConfigVersion {
+			return fmt.Errorf("breaking config version %q does not match lint config version %q", breakingConfigVersion, lintConfigVersion)
+		}
+		version = breakingConfigVersion
+	} else if breakingConfigVersion != "" {
+		version = breakingConfigVersion
+	} else if lintConfigVersion != "" {
+		version = lintConfigVersion
+	}
```
Comment (amckinney) [inline private/bufpkg/bufmodule/bufmodule.go:499]: We should return an error in these cases. We should force the caller of this function to provide consistent versions for the module's `LintConfig` and `BreakingConfig`, otherwise it's an internal error on our end since that should always be in our control.

We should also be able to guarantee that the `{Breaking,Lint}Config` version is _always_ non-empty if the `{Breaking,Lint}Config` is non-`nil`.

So in short:
* A version _must_ be set on both of the `LintConfig` and `BreakingConfig` types if they are non-`nil`.
* The version _must_ be equivalent between them if they are both non-`nil`.
* If one of the `{Breaking,Lint}Config` types is `nil`, the versions don't necessarily need to match, but this isn't actually possible in practice because there's always a default value for each one. If one of the configs is provided, both will be non-`nil`.

File: private/bufpkg/bufconfig/write.go (hunk 1 above)
Comment (amckinney) [reply private/bufpkg/bufconfig/write.go:270]: I see now, it's probably right that we should default to `v1beta1` since that's consistent with the configuration getters - we just need to update the documentation on `WriteConfigWithVersion` to say that the default is `v1beta`.

File: private/bufpkg/bufconfig/write.go (hunk 3 above)
Comment (doriable) [reply private/bufpkg/bufconfig/write.go:283]: Yeah -- that's fair. The validation happens in the module build also well, but maybe it makes sense to actually validate the versions here and then move the version inferences out of `bufmodule.go`?

File: private/buf/cmd/buf/buf_test.go (hunk 2 above)
Comment (doriable) [reply private/buf/cmd/buf/buf_test.go:1166]: Nice -- this works, so I'm happy about this :D

File: private/bufpkg/bufconfig/write.go [hunk 9]
```diff
@@ -273,30 +257,173 @@ func writeConfig(
...
+}
+
+func validateVersion(version string) error {
+	switch version {
+	case V1Version, V1Beta1Version:
+		return nil
+	default:
+		return fmt.Errorf("invalid config version %q provided", version)
+	}
+}
```
Comment (amckinney) [inline private/bufpkg/bufconfig/write.go:375]: nit: Move this function down so that it's adjacent to the other validation helper (at the bottom of the file),

File: private/bufpkg/bufconfig/write.go [hunk 10]
```diff
@@ -273,30 +257,173 @@ func writeConfig(
...
-			dependencyModuleReference.String(),
-		)
+	var lintConfigVersion string
+	if writeConfigOptions.lintConfig != nil {
+		config.Lint = writeConfigOptions.lintConfig
+		lintConfigVersion = version
+	}
+	if breakingConfigVersion != lintConfigVersion {
+		return fmt.Errorf("breaking config version %q does not match lint config version %q", breakingConfigVersion, lintConfigVersion)
+	}
```
Comment (amckinney) [inline private/bufpkg/bufconfig/write.go:287]: I still don't know if this is right.

We should validate that both `breakingConfigVersion` and `lintConfigVersion` are non-empty and set to the exact same value as the `version` provided by `writeConfigOptions.version`, if any.

Otherwise you'll end up in that situation where the `writeConfigOptions.version` would incorrectly override the `{Breaking,Lint}`Config version as discussed in [this](https://github.com/bufbuild/buf/pull/766#discussion_r764439767) comment.

File: private/bufpkg/bufmodule/module.go [hunk 11]
```diff
@@ -89,6 +89,10 @@ func newModuleForBucket(
 	if err != nil {
 		return nil, err
 	}
+	// if the module config has an identity, set the module identity
+	if moduleConfig.ModuleIdentity != nil {
+		options = append(options, ModuleWithModuleIdentity(moduleConfig.ModuleIdentity))
+	}
```
Comment (amckinney) [inline private/bufpkg/bufmodule/module.go:95]: This will override the `ModuleWithModuleIdentity[AndCommit]` option provided by the caller of `NewModuleForBucket`, if one was provided:

```go
// NewModuleForBucket returns a new Module. It attempts reads dependencies
// from a lock file in the read bucket.
func NewModuleForBucket(
	ctx context.Context,
	readBucket storage.ReadBucket,
	options ...ModuleOption,
) (Module, error) {
	return newModuleForBucket(ctx, readBucket, options...)
}
```

In practice, this might be OK since the `ModuleIdentity` found in the bucket _should_ always be equivalent to the option provided (i.e. the option is set with the module pin in the module cacher).

However, this is prone to error - if there's ever a reason that the caller of `NewModuleForBucket` would need to _override_ the `ModuleIdentity` that would otherwise be determined by the `buf.yaml` (this condition), then the caller's option will actually be overwritten with what's found in the bucket.

--

One solution out of this is to _prepend_ this option, which allows the other call-site provided option to take effect _after_ this one, hence taking precedence. This is also nuanced behavior though and feels kinda hacky, so we would need to aggressively include comments here if that's our only option. For example,

```go
if moduleConfig.ModuleIdentity != nil {
	// Explain why we're doing this.
	options = append([]ModuleOption{ModuleWithModuleIdentity(moduleConfig.ModuleIdentity)}, options...)
}
```

Alternatively, we could add an argument to the `newModule` constructor below so that it takes a positional `bufmodule.ModuleIdentity` parameter (which may be `nil`). This would be the default value, but it could be naturally overwritten by the option.

```go
// this should only be called by other newModule constructors
func newModule(
	ctx context.Context,
	// must only contain .proto files
	sourceReadBucket storage.ReadBucket,
	moduleIdentity ModuleIdentity,
	dependencyModulePins []bufmoduleref.ModulePin,
	documentation string,
	breakingConfig *bufbreakingconfig.Config,
	lintConfig *buflintconfig.Config,
	options ...ModuleOption,
) (_ *module, retErr error) {
	if err := bufmoduleref.ValidateModulePinsUniqueByIdentity(dependencyModulePins); err != nil {
		return nil, err
	}
	// we rely on this being sorted here
	bufmoduleref.SortModulePins(dependencyModulePins)
	module := &module{
		sourceReadBucket:     sourceReadBucket,
		moduleIdentity:       moduleIdentity,
		dependencyModulePins: dependencyModulePins,
		documentation:        documentation,
		breakingConfig:       breakingConfig,
		lintConfig:           lintConfig,
	}
	for _, option := range options {
		option(module)
	}
	return module, nil
}
```

Both aren't ideal, but I think I'd lean towards the latter option (updating the `newModule` signature) since it's a lot easier to follow. It will require passing in a `nil` argument in the `newModule` referenced in `newModuleForProto` which is a code smell, but we don't have the `ModuleName` in the Protobuf representation of `Module` - it's explicitly passed in via the [ModuleReader](https://github.com/bufbuild/buf/blob/0786cea8690ed81cdd1f5b31b9b5c0529ffdab74/private/bufpkg/bufapimodule/module_reader.go#L67). If we go this route, we should add a comment in-line explaining this at that call-site.

What are your thoughts?

File: private/bufpkg/bufconfig/write.go (hunk 10 above)
Comment (doriable) [reply private/bufpkg/bufconfig/write.go:287]: Ah yes... hmm... I thought I changed this -- maybe I only changed it in the `bufmodule.go`. Let me fix this, my bad!

File: private/bufpkg/bufmodule/module.go (hunk 11 above)
Comment (doriable) [reply private/bufpkg/bufmodule/module.go:95]: After some offline discussion (and some flipflopping, lol), the option we're going to go with is the second, adding a new parameter to `newModule`.

The crux of this discussion is around the intended behaviour for the build. The original intention was that any `ModuleIdentity` set by the call site _should not_ deviate from the `ModuleIdentity` from the config getter. And in practice, this is the case, since there could be a lot of side effects and breakages caused by overriding with a malformed module identity (including to the digest, etc.). That being said, there could be some use cases, such as testing, etc. where the module identity will need to be overridden, and thus we should always respect the value provided by the call site.

Thus, by adding a new parameter, we can explicitly set the module identity when building via the bucket (and keep it `nil` via other build paths), and then still respect the option override.

File: private/bufpkg/bufmodule/module.go [hunk 12]
```diff
@@ -65,6 +65,7 @@ func newModuleForProto(
 		ctx,
 		readWriteBucket,
 		dependencyModulePins,
+		nil,
```
Comment (amckinney) [inline private/bufpkg/bufmodule/module.go:68]: Let's include that comment here - that the Module Identity isn't part of the Protobuf definition, but it's expected to be provided by the layer above (e.g. the `ModuleReader`).

File: private/bufpkg/bufconfig/write.go [hunk 13]
```diff
@@ -273,30 +257,170 @@ func writeConfig(
...
+			return fmt.Errorf("invalid version found for breaking config: %q", breakingConfig.Version)
+		}
+		config.Breaking = breakingConfig
 	}
-	for _, dependencyModuleReference := range writeConfigOptions.dependencyModuleReferences {
-		externalConfigV1.Deps = append(
-			externalConfigV1.Deps,
-			dependencyModuleReference.String(),
-		)
+	var lintConfigVersion string
```
[the remark this answers is not in the document]
Comment (doriable) [reply private/bufpkg/bufconfig/write.go:283]: 🤦🏼‍♀️

File: private/bufpkg/bufconfig/write_test.go [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+		writeConfigOptions := []WriteConfigOption{
+			WriteConfigWithVersion(V1Beta1Version),
+			WriteConfigWithLintConfig(&buflintconfig.Config{
+				Version: V1Version,
+				Use:     []string{"DEFAULT"},
+			}),
+		}
+		err = WriteConfig(context.Background(), readWriteBucket, writeConfigOptions...)
+		require.Error(t, err)
+		require.Equal(t, err.Error(), `invalid version found for lint config: "v1"`)
```
Comment (amckinney) [inline private/bufpkg/bufconfig/write_test.go:94]: nit: we should say what we expect the version to be - otherwise the error message is confusing (`v1` is a valid version).

File: private/bufpkg/bufconfig/write.go [hunk 15]
```diff
@@ -273,30 +257,173 @@ func writeConfig(
...
+			Deps:     dependencies,
+			Breaking: externalBreakingConfig,
+			Lint:     externalLintConfig,
+		}
+		var buffer bytes.Buffer
+		encoder := yaml.NewEncoder(&buffer)
+		encoder.SetIndent(2)
+		if err := encoder.Encode(externalConfig); err != nil {
+			return err
+		}
```
Comment (amckinney) [inline private/bufpkg/bufconfig/write.go:347]: Discussed offline - 

This will produce non-deterministic results because the YAML structure contains a `map[string][]string` nested within it.

Right now, this isn't actually user-facing - the only time we actually write these potentially non-deterministic values is in the users' local module cache. It's definitely not ideal, but we can follow-up with a fix after this PR is merged (i.e. using the `yaml.Node` API in `yaml.v3`).

What matters most is that the digest algorithm is deterministic and the `buf.yaml` can be successfully round-tripped without losing any information (just sometimes in a different order, such as with `ignore_only`).

We will need a fix in these representations before encoding the `buf.yaml` is formally complete.

File: private/bufpkg/bufmodule/bufmodulecache/bufmodulecache_test.go [hunk 16]
```diff
@@ -269,6 +267,42 @@ func TestModuleReaderCacherWithDocumentation(t *testing.T) {
...
+	err = moduleCacher.PutModule(
+		context.Background(),
+		modulePin,
+		module,
+	)
+	require.NoError(t, err)
+	module, err = moduleCacher.GetModule(ctx, modulePin)
+	require.NoError(t, err)
+	readWriteBucket := storagemem.NewReadWriteBucket()
+	require.NoError(t, bufmodule.ModuleToBucket(ctx, module, readWriteBucket))
```
Comment (amckinney) [inline private/bufpkg/bufmodule/bufmodulecache/bufmodulecache_test.go:299]: This test is valuable, but is there anything else we can add to it with respect to verifying the configuration is round-trippable?

Can we verify that the module we put is exactly the same as the module we get out? Right now it looks like we're only doing so for the documentation below.

File: private/bufpkg/bufconfig/write_test.go (hunk 14 above)
Comment (doriable) [reply private/bufpkg/bufconfig/write_test.go:94]: Ah yeah, that's a good call.

File: private/bufpkg/bufconfig/write.go (hunk 15 above)
Comment (doriable) [reply private/bufpkg/bufconfig/write.go:347]: +1, thanks for comment :) I am currently working on summarizing the requirements for having the config be round-trippable in a stand-alone issue, does it makes sense to bundle that as a strict requirement for the encoding `buf.yaml` issue or should it be its own thing?

File: private/bufpkg/bufmodule/bufmodulecache/bufmodulecache_test.go (hunk 16 above)
Comment (doriable) [reply private/bufpkg/bufmodule/bufmodulecache/bufmodulecache_test.go:300]: Yep, just added to the test case to check against the `bufmoduletesting.TestModuleConfiguration` config.

**Review by amckinney — APPROVED** [review]
Looks good!

File: private/buf/cmd/buf/command/config/configinit/configinit.go [hunk 17]
```diff
@@ -126,6 +128,26 @@ func run(
...
 	}
+	// Need to include the default version (v1), lint config, and breaking config.
+	version := bufconfig.V1Version
+	writeConfigOptions = append(
+		writeConfigOptions,
+		bufconfig.WriteConfigWithVersion(version),
+	)
+	writeConfigOptions = append(
+		writeConfigOptions,
+		bufconfig.WriteConfigWithBreakingConfig(&bufbreakingconfig.Config{
```
Comment (bufdev) [inline private/buf/cmd/buf/command/config/configinit/configinit.go:139]: Newline after ( and after final }

File: private/buf/cmd/buf/command/config/configinit/configinit.go [hunk 18]
```diff
@@ -126,6 +128,26 @@ func run(
...
+			Version: version,
+			Use:     []string{"FILE"},
+		}),
+	)
+	writeConfigOptions = append(
+		writeConfigOptions,
+		bufconfig.WriteConfigWithLintConfig(&buflintconfig.Config{
```
Comment (bufdev) [inline private/buf/cmd/buf/command/config/configinit/configinit.go:146]: Newline after ( and after final }

File: private/bufpkg/bufconfig/write.go [hunk 19]
```diff
@@ -273,30 +257,173 @@ func writeConfig(
...
 	}
-	if !writeConfigOptions.documentationComments && writeConfigOptions.uncomment {
-		return errors.New("cannot set uncomment without documentationComments for WriteConfig")
+	if err := validateWriteConfigOptions(writeConfigOptions); err != nil {
+		return err
 	}
-	if writeConfigOptions.moduleIdentity == nil && len(writeConfigOptions.dependencyModuleReferences) > 0 {
-		return errors.New("cannot set deps without a name for WriteConfig")
+	// This is the same default as the bufconfig getters.
+	version := V1Beta1Version
```
Comment (bufdev) [inline private/bufpkg/bufconfig/write.go:264]: This seems wrong for the default value for write, are we sure about this?

File: private/bufpkg/bufconfig/write.go [hunk 20]
```diff
@@ -273,30 +257,173 @@ func writeConfig(
...
-	externalConfigV1.Lint.Use = defaultLintIDs
-	externalConfigV1.Breaking.Use = defaultBreakingIDs
-	if writeConfigOptions.moduleIdentity != nil {
-		externalConfigV1.Name = writeConfigOptions.moduleIdentity.IdentityString()
+	var breakingConfigVersion string
+	breakingConfig := writeConfigOptions.breakingConfig
+	if breakingConfig != nil {
+		breakingConfigVersion = breakingConfig.Version
+		if breakingConfigVersion != version {
+			return fmt.Errorf("version %q found for breaking config, does not match top level config version: %q", breakingConfigVersion, version)
```
Comment (bufdev) [inline private/bufpkg/bufconfig/write.go:280]: This worries me - I'm not sure what to do about it but it seems like an API flaw that we can have mismatched versioms

File: private/bufpkg/bufconfig/write.go [hunk 21]
```diff
@@ -273,30 +257,173 @@ func writeConfig(
...
+			externalLintConfig = buflintconfig.ExternalConfigV1Beta1ForConfig(lintConfig)
+		}
+		externalConfig := ExternalConfigV1Beta1{
+			Name:     name,
+			Version:  config.Version,
+			Deps:     dependencies,
+			Breaking: externalBreakingConfig,
+			Lint:     externalLintConfig,
+		}
+		var buffer bytes.Buffer
```
Comment (bufdev) [inline private/bufpkg/bufconfig/write.go:342]: We do `buffer := bytes.NewBuffer(nil)` in the rest of the codebase

File: private/bufpkg/bufconfig/write.go [hunk 22]
```diff
@@ -273,30 +257,173 @@ func writeConfig(
...
+		name = config.ModuleIdentity.IdentityString()
+	}
+	if name == "" {
+		name = exampleName
+		nameUnset = true
 	}
+	var dependenciesUnset bool
+	if len(dependencies) == 0 {
+		dependencies = []string{
+			"buf.build/acme/petapis",
```
Comment (bufdev) [inline private/bufpkg/bufconfig/write.go:401]: Why were these removed from being consts?

File: private/bufpkg/bufconfig/write_test.go [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+func TestWriteConfigSuccess(t *testing.T) {
+	storageosProvider := storageos.NewProvider(storageos.ProviderWithSymlinks())
+	readWriteBucket, err := storageosProvider.NewReadWriteBucket(t.TempDir())
+	require.NoError(t, err)
+	writeConfigOptions := []WriteConfigOption{
+		WriteConfigWithVersion(V1Version),
+		WriteConfigWithBreakingConfig(&bufbreakingconfig.Config{
```
Comment (bufdev) [inline private/bufpkg/bufconfig/write_test.go:34]: newline

File: private/bufpkg/bufconfig/write_test.go [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+			Version: V1Version,
+			Use:     []string{"FILE"},
+		}),
+		WriteConfigWithLintConfig(&buflintconfig.Config{
```
Comment (bufdev) [inline private/bufpkg/bufconfig/write_test.go:38]: newline

File: private/bufpkg/bufmodule/bufmodule.go [hunk 25]
```diff
@@ -472,7 +473,36 @@ func ModuleToBucket(
 			return err
 		}
 	}
-	return bufmoduleref.PutDependencyModulePinsToBucket(ctx, writeBucket, module.DependencyModulePins())
+	if err := bufmoduleref.PutDependencyModulePinsToBucket(ctx, writeBucket, module.DependencyModulePins()); err != nil {
+		return err
+	}
+	// This is the default version created by bufconfig getters. The versions should be the
+	// same across lint and breaking configs.
+	version := bufconfig.V1Beta1Version
```
Comment (bufdev) [inline private/bufpkg/bufmodule/bufmodule.go:481]: Same question as before

File: private/bufpkg/bufmodule/bufmodule.go [hunk 26]
```diff
@@ -472,7 +473,36 @@ func ModuleToBucket(
...
+	var breakingConfigVersion string
+	if module.BreakingConfig() != nil {
+		breakingConfigVersion = module.BreakingConfig().Version
```
Comment (bufdev) [inline private/bufpkg/bufmodule/bufmodule.go:484]: Why are we exposing `Version` at all? This feels like an external concern, whereas this should be abstracted away for internal types.

File: private/bufpkg/bufconfig/write.go (hunk 20 above)
Comment (doriable) [reply private/bufpkg/bufconfig/write.go:280]: Yeah, this should not be a code path that we ever hit and would be an API level problem, but I do think it's important to surface this error if we are attempting to write invalid config.

File: private/bufpkg/bufconfig/write.go (hunk 22 above)
Comment (doriable) [reply private/bufpkg/bufconfig/write.go:401]: I moved them here since they are only being used for the template rendering, which is only used by `configinit` and I figured we should reduce the number globals. But I can move them back if that is preferred.

File: private/bufpkg/bufconfig/write.go (hunk 19 above)
Comment (doriable) [reply private/bufpkg/bufconfig/write.go:264]: Yeah, this is consistent with the behaviour for when we go to get configs, if none are found, we default to `v1beta1` right now: https://github.com/bufbuild/buf/blob/f4fd40a47101848d722041ee6bfec433621bf1fe/private/bufpkg/bufconfig/get.go#L48-L50

There is a note to change this when we move to `v1`.

File: private/bufpkg/bufmodule/bufmodule.go (hunk 25 above)
Comment (doriable) [reply private/bufpkg/bufmodule/bufmodule.go:481]: Addressed above.

File: private/bufpkg/bufmodule/bufmodule.go (hunk 26 above)
Comment (doriable) [reply private/bufpkg/bufmodule/bufmodule.go:484]: This is based because the `Config` structure we store needs to reflect the external type for the digest. The config we store alongside the module needs to reflect changes in the configuration being sent by the user so it can be updated accordingly.
